man/npcd.8.in
.\" In .TH, FOO should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.\"
.\" This template provided by Tom Christiansen <tchrist@jhereg.perl.com>.
.\" 
.TH NPCD 8
.SH NAME
npcd \- Nagios Performance C Daemon
.SH SYNOPSIS
\fBnpcd\fR [ \fB\-d | \--daemon\fR ] ( \fB\-f | \--config\fR ) \fI<configfile>
.SH DESCRIPTION
The \fBnpcd\fR is a daemon processing nagios performance data.
.PP
For more information you can visit pnp4nagios website:
.IP
http://docs.pnp4nagios.org/
.PP
and SourceForge project:
.IP
http://pnp4nagios.sourceforge.net
.SH OPTIONS
.TP
\fB\-d, \--daemon
Run \fBnpcd\fR as daemon in the background.
.TP
\fB\-f, \--config\fR \fI<configfile>
Path to the \fBnpcd\fR config file (typically @sysconfdir@/npcd.cfg).
.SH "RETURN VALUE"
On success starting \fBnpcd\fR 0 is returned.
.SH ERRORS
On error starting \fBnpcd\fR 1 is returned.
.SH FILES
.br
.nf
\fI@sysconfdir@/npcd.cfg\fR	default config file
.SH AUTHORS
Joerg Linge <pitchfork@pnp4nagios.org>, 
Hendrik Baecker <andurin@process-zero.de>


ChangeLog
**pnp-0.6.13 ??/??/2011**
  * Feature: New option --ignore-hosts added to check_pnp_rrds.pl ( by Jochen Bern )

**pnp-0.6.12 04/22/2011**
  * Feature: mod_gearman support added
  * Feature: rrd_convert.pl is now able to convert all RRDs from RRD_STORAGE_TYPE=SINGLE to RRD_STORAGE_TYPE=MULTIPLE
  * Feature: New template check_gearman.php
  * Feature: Install process_perfdata.cfg and npcd.cfg by default
  * Bugfix:  rrd_convert.pl is now able to parse xml dumps created by rrdtool 1.4.x  
  * Bugfix:  process_perfdata.pl default timeout value set to 15 seconds
  * Info: Version used by OMD-0.48 [[http://omdistro.org|OMD]]

**pnp-0.6.11 01/15/2011**
  * Bugfix: urldecoding fixed 
  * Bugfix: Zoom in/out is working again ( Reported by Thorben Soehl )
  * Featue: npcd.cfg -  New option perfdata_file_processing_interval used by npcdmod
  * Info: Version used by OMD-0.46 [[http://omdistro.org|OMD]]

**pnp-0.6.10 12/15/2010**

  * Feature: Add RRDTool Option --only-graph if graph height is below 32px to create thumbnails 
  * Feature: RRDTool Option --width and --height is now allowed in templates
  * Feature: RRDTool DS Type for UOM of "c" changed from COUNTER to DERIVE   
  * Feature: Pass query string from special controller to image controller ( Matthew Garrett ) 
  * Feature: Authorisation against [[http://mathias-kettner.de/checkmk_livestatus.html|mk_livestatus]] API added
  * Feature: Sample nginx webserver config added ( by evax@users.sourceforge.net )
  * Feature: Kohana backport to support PHP 5.1.6 ( Kudos to Andreas Ericsson )
  * Bugfix: Sort list of special templates
  * Bugfix: Urlencode hostname and service description  ( Reported by Yannick ) 
  * Bugfix: corrected warning/critical thresholds in ticker/alerter functions

**pnp-0.6.7 09/27/2010**

  * Bugfix: Page config parser fix (Beau Gunderson) 
  * Bugfix: Zoom window size fixed (Report by Rudolf Labuschagne) 
  * Bugfix: Fixed undefined offset while using 'ds_name' in templates (Reported by Vladimir Bilik)
  * Bugfix: Npcdmod respects process_perf_data option used in hosts and services definitions ( Reported by Wolfgang Barth )
  * Template: New Template check_nagiostats.php used with check_nagiostats written by Jochen Bern 

**pnp-0.6.6 08/07/2010**

  * Bugfix: Fixed max amount of graphs per template 
  * Bugfix: Autodetect PNP base URL
  * Bugfix: Too short npcdmod perfdata_template to take perfdata + overhead, increased +1024byte
  * Bugfix: Ignore files in var/perfdata and check for empty directories
  * Bugfix: Reducing memory usage while parsing page config (Laurent Freval)

**pnp-0.6.5 07/09/2010**

  * Feature: Special Templates are back [[tpl_special]]
  * Feature: New rrdtool helper functions makes template design easier [[tpl_helper]]
  * Feature: config.php -> 'recursive_template_search' is enabled by default
  * Feature: config.php -> 'template_dirs' is now an array of directorys to search for PNP templates

**pnp-0.6.4 06/03/2010**

  * Update: jQuery Update to 1.4.2
  * Update: jQuery-ui Update to 1.8
  * Feature: New configure Option --with-base-url
  * Template: New template check_ntp_time.php (Mathias Kettner)
  * Feature: New i18n files for fr_FR (Yannig Parre)
  * Feature: New jQuery Theme 'multisite'

**pnp-0.6.3 03/16/2010**

  * Feature: New helper script libexec/rrd_convert.pl -> [[rrd_convert]]
  * Bugfix: Ignore old XML files while building the service list
  * Template: New template check_hpasm.php
  * Bugfix: Installer now checks for json_decode()
  * Workaround: Allow "trailing unfilled semicolons". Workaround for nsclient++
  * Template: Updates for check_openmanage.php, check_hp_bladecenter.php and check_dell_baldecenter.php ( Trond Hasle Amundsen )"

**pnp-0.6.2 12/23/2009**

  * Feature: XML_WRITE_DELAY option added to process_perfdata.cfg as suggested by Mathias Kettner
  * Feature: New template integer.php 
  * Update: FPDI update to 1.3.1
  * Feature: PNP will now work with [[http://http://www.lighttpd.net|lighttpd]] and php-cgi
  * Template: check_mk-ps.perf.php added ( by Mathias Kettner )
  * Feature: PNP will now work without mod_rewrite -> [[webfe]] 
  * Bugfix: Wrong pdf link used on site 'pages' and 'basket'
  * Bugfix: Incorrect group permissions on spool directory

**pnp-0.6.1 11/22/2009**

  * Feature: RRD heartbeat per check_command -> [[tpl_custom]] 
  * Feature: New config.php option pdf_graph_opt 
  * Feature: Recognize the 'background_pdf' option in page definitions -> [[pages]]
  * Feature: Recognize the 'source' option in page definitions -> [[pages]]
  * Feature: Array $TIMERANGE now available for templates -> [[timeranges]]
  * Bugfix:  ./configure --sysconfdir no longer ignored
  * Feature: Store internal runtime statistics on a per minute base 
  * Feature: Added two widgets views/widget_menu.php and views/widget_graph.php 

**pnp-0.6.0 10/30/2009**

  * Webfrontend based on [[http://www.kohanaphp.com|Kohana]]
  * Webfrontend based on [[http://jqueryui.com/themeroller/|jQuery Themes]]
  * Javascript-functions using [[http://jquery.com/|jQuery]] plugins  
  * process_perfdata.pl will be able to use one RRD database per datasource
  * improved installer. Specification of directory layouts using --with-layout
  * RRDtool errors are now displayed as images. no more missing images
  * PNP templates cannot overwrite internal variables anymore
  * PNP templates of version 0.4.x can still be used
  * PDF functions recoded
  * Template default.php optimized
  * Export from RRD databases into XML, CSV and JSON format using the RRDtool "xport" function
  * Page functions recoded
  * Error pages links to online FAQ
  * Mouseover Popup in Nagios frontend via jQuery.clueTip plugin
  * Full support of rrdcached


src/npcd.c
/* Copyright (C) 2007-2009 Hendrik Baecker <andurin@process-zero.de>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License version 2 as
* published by the Free Software Foundation;
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "../include/config.h"
#include "../include/pnp.h"

typedef void (*sighandler_t)(int);

void *processfile(void *);
static void *exit_handler_mem(void *);
extern int process_arguments(int, char **);
extern void process_configfile(char *);
extern void check_sig(int);
extern int prepare_vars(void);
extern int drop_privileges(char *, char *);
extern sighandler_t handle_signal(int, sighandler_t);
extern int is_file(const struct dirent *d);
extern int check_needed_config_options();

double getload(int);

static int thread_counter = 0;
int max_threads = 5;
int daemon_mode = FALSE;
int use_syslog = TRUE;
int loglevel = 0;
int max_logfile_size = 10485760; /* default 10Mbyte */
int use_load_threshold = FALSE;
int we_should_stop = FALSE;
int sleeptime = 15;
int identmyself = TRUE;
double load_threshold = 10.0;

extern int sighup_detected;

char *command, *command_args, *user, *group, *pidfile;
char *macro_x[CONFIG_OPT_COUNT];
char *log_file, *log_type;
char *config_file = NULL;

const char *directory = NULL;
const char progname[5] = "npcd";

static void start_daemon(const char *log_name, int facility) {
	int i;
	pid_t pid;

	/* Kill parent after for to get a waise */
	if ((pid = fork()) != 0)
		exit(EXIT_SUCCESS);

	/* Get this waise to sessionleader */
	if (setsid() < 0) {
		printf("%s could not get sessionleader\n", log_name);
		exit(EXIT_FAILURE);
	}

	/* Ignore SIGHUP */
	handle_signal(SIGHUP, SIG_IGN);

	/* terminate child */
	if ((pid = fork()) != 0)
		exit(EXIT_SUCCESS);

	/* for core dump handling and better unmounting behavior */
	if (chdir("/") != 0)
		exit(EXIT_FAILURE);

	/* change umask to defined value - be independet from parent umask */
	umask(002);

	/* close all possible file handles */
	for (i = sysconf(_SC_OPEN_MAX); i > 0; i--)
		close(i);

	/* to hear the daemon you are calling... use syslog */
	if (use_syslog == TRUE && log_name != NULL)
		openlog(log_name, LOG_PID | LOG_CONS | LOG_NDELAY, facility);

        /* close existing stdin, stdout, stderr */
        close(0);
        close(1);
        close(2);

        /* re-open stdin, stdout, stderr with known values */
        open("/dev/null",O_RDONLY);
        open("/dev/null",O_WRONLY);
        open("/dev/null",O_WRONLY);

}

int main(int argc, char **argv) {

	int i = 0;
	int filecounter = 0, pthread_ret = 0;
	double load = 0.0;
	char buffer[MAX_LOGMESSAGE_SIZE];

	FILE *fppid = NULL;

	struct dirent **namelist;

	if (process_arguments(argc, argv) == EXIT_FAILURE)
		exit(EXIT_FAILURE);

	process_configfile(config_file);

	if (loglevel == -1) {
		printf("DEBUG: Config File = %s\n", config_file);
		printf("CONFIG_OPT_LOGTYPE = %s\n", macro_x[CONFIG_OPT_LOGTYPE]);
		printf("CONFIG_OPT_LOGFILE = %s\n", macro_x[CONFIG_OPT_LOGFILE]);
		printf("CONFIG_OPT_LOGFILESIZE = %s\n", macro_x[CONFIG_OPT_LOGFILESIZE]);
		printf("CONFIG_OPT_LOGLEVEL = %s\n", macro_x[CONFIG_OPT_LOGLEVEL]);
		printf("CONFIG_OPT_SCANDIR = %s\n", macro_x[CONFIG_OPT_SCANDIR]);
		printf("CONFIG_OPT_RUNCMD = %s\n", macro_x[CONFIG_OPT_RUNCMD]);
		printf("CONFIG_OPT_RUNCMD_ARG = %s\n", macro_x[CONFIG_OPT_RUNCMD_ARG]);
		printf("CONFIG_OPT_MAXTHREADS = %s\n", macro_x[CONFIG_OPT_MAXTHREADS]);
		printf("CONFIG_OPT_LOAD = %s\n", macro_x[CONFIG_OPT_LOAD]);
		printf("CONFIG_OPT_USER = %s\n", macro_x[CONFIG_OPT_USER]);
		printf("CONFIG_OPT_GROUP = %s\n", macro_x[CONFIG_OPT_GROUP]);
		printf("CONFIG_OPT_PIDFILE = %s\n", macro_x[CONFIG_OPT_PIDFILE]);
		printf("CONFIG_OPT_SLEEPTIME = %s\n", macro_x[CONFIG_OPT_SLEEPTIME]);
		printf("CONFIG_OPT_IDENTMYSELF = %s\n", macro_x[CONFIG_OPT_IDENTMYSELF]);
		printf("---------------------------\n");
		if (check_needed_config_options() != 0) {
			printf("There is an Error! Exiting...\n");
			exit(EXIT_FAILURE);
		}
	}

	if (prepare_vars() != 0)
		exit(EXIT_FAILURE);
	if (loglevel == -1)
		printf("DEBUG: load_threshold is %s - ('%f')\n",
				use_load_threshold ? "enabled" : "disabled", load_threshold);

	pthread_t th[max_threads];
	for (i=0;i<max_threads;i++){
		th[i] = (pthread_t) NULL;
	}
	i = 0;

	/* Nice point for another function to set
	 * the internal vars from macro_x[] */

	/* Start in Daemon Mode or in foreground? */
	if (daemon_mode == TRUE)
		start_daemon("NPCD", LOG_LOCAL0);

	else if (use_syslog)
		openlog("NPCD", LOG_PID | LOG_CONS | LOG_NDELAY, LOG_LOCAL0);

	/* Create PID File or exit on failure */
	if (daemon_mode == TRUE && sighup_detected == FALSE) {
		fppid = fopen(pidfile, "w");

		if (fppid == NULL) {
			printf("Could not open pidfile '%s': %s\n", pidfile,
					strerror(errno));
			exit(EXIT_FAILURE);
		} else {
			fprintf(fppid, "%d", getpid());
			fclose(fppid);
		}
	}

	/* Try to drop the privileges */
	if (drop_privileges(user, group) == EXIT_FAILURE)
		exit(EXIT_FAILURE);

	snprintf(buffer, sizeof(buffer) - 1,
			"%s Daemon (%s) started with PID=%d\n", progname, PACKAGE_VERSION,
			getpid());
	LOG(0, buffer);
	snprintf(buffer, sizeof(buffer) - 1,
			"Please have a look at '%s -V' to get license information\n",
			progname);
	LOG(0, buffer);

	//sigemptyset();
	handle_signal(SIGINT, check_sig);
	handle_signal(SIGHUP, check_sig);
	handle_signal(SIGTERM, check_sig);

	snprintf(buffer, sizeof(buffer) - 1,
			"HINT: load_threshold is %s - ('%f')\n",
			use_load_threshold ? "enabled" : "disabled", load_threshold);
	LOG(0, buffer);

	/* beginn main loop */
	while (1) {

		/* read directory with perfdata files */
		if (chdir(directory) != 0)
			exit(EXIT_FAILURE);

		/* is_file() filter may cause trouble on some systems
		 * like Solaris or HP-UX that don't have a d-type
		 * member in struct dirent
		 */
		/* #ifdef HAVE_STRUCT_DIRENT_D_TYPE
		 if ( ( filecounter = scandir( directory, &namelist, is_file, alphasort ) ) < 0 ) {
		 #else */
		if ((filecounter = scandir(directory, &namelist, 0, alphasort)) < 0) {
			/* #endif */
			snprintf(buffer, sizeof(buffer) - 1,
					"Error while get file list from spooldir (%s) - %s\n",
					directory, strerror(errno));
			LOG(0, buffer);
			snprintf(buffer, sizeof(buffer) - 1, "Exiting...\n");
			LOG(0, buffer);

			if (daemon_mode != TRUE)
				printf("Error while get file list from spooldir (%s) - %s\n",
						directory, strerror(errno));
			break;
		}

		snprintf(buffer, sizeof(buffer) - 1, "Found %d files in %s\n",
				filecounter, directory);
		LOG(2, buffer);

		for (i = 0, namelist; i < filecounter; i++) {

#ifdef HAVE_GETLOADAVG
			if (use_load_threshold == TRUE) {
				load = getload(1);
				snprintf(buffer, sizeof(buffer) - 1, "DEBUG: load %f/%f\n",
						load, load_threshold);
				LOG(2, buffer);
			}

			if (use_load_threshold && (load > load_threshold)) {

				snprintf(buffer, sizeof(buffer) - 1,
						"WARN: MAX load reached: load %f/%f at i=%d", load,
						load_threshold, i);
				LOG(0, buffer);

				if (i > 0)
					i--;
				sleep(sleeptime);
				continue;
			}
#endif

			snprintf(buffer, sizeof(buffer) - 1,
					"ThreadCounter %d/%d File is %s\n", thread_counter,
					max_threads, namelist[i]->d_name);
			LOG(2, buffer);

			struct stat attribute;

			if (stat(namelist[i]->d_name, &attribute) == -1) {
				LOG(0, "Error while getting file status");
				break;
			}

			if (strstr((namelist[i]->d_name), "-PID-") != NULL) {
				snprintf(
						buffer,
						sizeof(buffer) - 1,
						"File '%s' is an already in process PNP file. Leaving it untouched.\n",
						namelist[i]->d_name);

				LOG(1, buffer);
				continue;
			}

			if (S_ISREG(attribute.st_mode)) {
				snprintf(buffer, sizeof(buffer) - 1, "Regular File: %s\n",
						namelist[i]->d_name);
				LOG(2, buffer);

				/* only start new threads if the max_thread config option is not reached */
				if (thread_counter < max_threads && we_should_stop == FALSE) {

					if ((pthread_ret = pthread_create(&th[thread_counter],
							NULL, processfile, namelist[i]->d_name)) != 0) {
						snprintf(buffer, sizeof(buffer) - 1,
							"Could not create thread... exiting with error '%s'\n", strerror(errno));
						LOG(0, buffer);
						exit(EXIT_FAILURE);
					}

					snprintf(buffer, sizeof(buffer) - 1,
							"A thread was started on thread_counter = %d\n",
							thread_counter);
					LOG(2, buffer);

					thread_counter++;

				}

				else if (we_should_stop == TRUE)
					break;

				else {

					snprintf(
							buffer,
							sizeof(buffer) - 1,
							"WARN: MAX Thread reached: %s comes later with ThreadCounter: %d\n",
							namelist[i]->d_name, thread_counter);
					LOG(2, buffer);

					i--;

					for (thread_counter = thread_counter; thread_counter > 0; thread_counter--) {
						snprintf(buffer, sizeof(buffer) - 1,
								"DEBUG: Will wait for th['%d']\n",
								thread_counter - 1);
						LOG(2, buffer);
						pthread_join(th[thread_counter - 1], NULL);
					}
				}
			}
		}

		if (thread_counter > 0) {
			/* Wait for open threads before working on the next run */
			snprintf(buffer, sizeof(buffer) - 1,
					"Have to wait: Filecounter = %d - thread_counter = %d\n",
					filecounter - 2, thread_counter);
			LOG(2, buffer);

			for (thread_counter = thread_counter; thread_counter > 0; thread_counter--)
				pthread_join(th[thread_counter - 1], NULL);
		}

		if (we_should_stop == TRUE)
			break;

		for (i = 0, namelist; i < filecounter; i++) {
			free(namelist[i]);
		}

		free(namelist);

		snprintf(buffer, sizeof(buffer) - 1,
				"No more files to process... waiting for %d seconds\n",
				sleeptime);
		LOG(1, buffer);

		sleep(sleeptime);

	}

	snprintf(buffer, sizeof(buffer) - 1, "Daemon ended. PID was '%d'\n",
			getpid());
	LOG(0, buffer);

	if (use_syslog)
		closelog();
	return EXIT_SUCCESS;
}

/* **************************************************************
 *
 * Function to parse and check the commandline arguments
 *
 * *************************************************************/

int process_arguments(int argc, char **argv) {
	int c;
	int error = FALSE;
	int display_license = FALSE;
	int display_help = FALSE;

#ifdef HAVE_GETOPT_H
	int option_index = 0;
	static struct option long_options[] = { { "help", no_argument, 0, 'h' }, {
			"version", no_argument, 0, 'V' },
			{ "license", no_argument, 0, 'V' },
			{ "daemon", no_argument, 0, 'd' }, { "config", required_argument,
					0, 'f' }, { 0, 0, 0, 0 } };
#endif

	/* make sure we have the correct number of command line arguments */
	if (argc < 2)
		error = TRUE;

	while (1) {

#ifdef HAVE_GETOPT_H
		c = getopt_long(argc, argv, "+hVdf:", long_options, &option_index);
#else
		c = getopt(argc, argv, "+hVdf:");
#endif

		if (c == -1 || c == EOF)
			break;

		switch (c) {

		case '?': /* usage */
		case 'h':
			display_help = TRUE;
			break;

		case 'V': /* version */
			printf("%s %s - $Revision: 637 $\n\n", progname, PACKAGE_VERSION);
			display_license = TRUE;
			break;

		case 'd': /* run in daemon mode */
			daemon_mode = TRUE;
			break;

		case 'f': /* config file */
			if (optarg != NULL)
				config_file = optarg;
			break;

		default:
			break;
		}

	}

	if (display_license == TRUE) {

		printf(
				"This program is free software; you can redistribute it and/or modify\n");
		printf(
				"it under the terms of the GNU General Public License version 2 as\n");
		printf("published by the Free Software Foundation.\n\n");
		printf(
				"This program is distributed in the hope that it will be useful,\n");
		printf(
				"but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
		printf(
				"MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
		printf("GNU General Public License for more details.\n\n");
		printf(
				"You should have received a copy of the GNU General Public License\n");
		printf("along with this program; if not, write to the Free Software\n");
		printf(
				"Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA\n\n");

		exit(EXIT_SUCCESS);
	}

	/* if there are no command line options (or if we encountered an error), print usage */
	if (error == TRUE || display_help == TRUE) {
		printf("\nUsage: %s -f <configfile> [-d] \n", argv[0]);
		printf("\n");
		printf("Options:\n");
		printf("\n");
		printf("  -d | --daemon \n");
		printf("\t\tRun as daemon in background\n");
		printf("\n");
		printf("  -f | --config \n");
		printf("\t\tPath to config file\n");
		printf("\n");
		printf(
				"Visit the Website at http://sourceforge.net/projects/pnp4nagios/ for bug fixes, \n");
		printf("new releases, online documentation, FAQs, Mailinglists.\n");
		printf("\n");

		exit(EXIT_FAILURE);
	}
	return EXIT_SUCCESS;
}

/********************************************************************
 *								    *
 * processfile - this is the function for each thread		    *
 *								    *
 ********************************************************************/

void * processfile(void *filename) {

	char *file = (char *) filename;
	char command_line[MAX_COMMANDLINE_LENGTH];
	char buffer[MAX_LOGMESSAGE_SIZE];
	int result = 0;
	FILE *proc;

	snprintf(command_line, sizeof(command_line), "%s %s %s %s/%s", command,
			identmyself ? "-n" : "\b", command_args, directory, file);

	pthread_cleanup_push((void *) &exit_handler_mem, file);

		snprintf(buffer, sizeof(buffer) - 1,
				"Processing file %s with ID %ld - going to exec %s\n", file,
				pthread_self(), command_line);
		LOG(2, buffer);

		snprintf(buffer, sizeof(buffer) - 1, "Processing file '%s'\n", file);
		LOG(1, buffer);

		if ((proc = popen(command_line, "r")) != NULL)
			result = pclose(proc);

		result >>= 8;

		if (result != 0) {
			snprintf(buffer, sizeof(buffer) - 1,
					"ERROR: Executed command exits with return code '%d'\n",
					result);
			LOG(0, buffer);

			snprintf(buffer, sizeof(buffer) - 1,
					"ERROR: Command line was '%s'\n", command_line);
			LOG(0, buffer);

			we_should_stop = FALSE;
		}

		if (loglevel == -1)
			sleep(2);

		pthread_cleanup_pop(1);
	pthread_exit((void *) pthread_self());
}

/* ******************************************************************
 *                                                                  *
 * processfile - this is the function for each thread               *
 *                                                                  *
 * ******************************************************************/

static void *exit_handler_mem(void * arg) {
	// syslog( LOG_NOTICE, "Will now clean up thread %ld\n",pthread_self());
	//if (thread_counter > 0)
	//thread_counter--;
	return 0;
}

#ifdef HAVE_GETLOADAVG
double getload(int which_sample) {

	double loadavg[3];
	char buffer[MAX_LOGMESSAGE_SIZE];

	if (which_sample == 1) {
		which_sample = 0;
	} else if (which_sample == 5) {
		which_sample = 1;
	} else if (which_sample == 15) {
		which_sample = 2;
	} else {
		snprintf(buffer, sizeof(buffer) - 1,
				"Invalid load sample %d - allowed is 1,5,15\n", which_sample);
		LOG(0, buffer);

		return -1;
	}
	getloadavg(loadavg, 3);
	return loadavg[which_sample];
}
#endif
